kernel/src/actions/crc.rs

```diff
@@ -94,14 +93,46 @@ pub(crate) struct DeletedRecordCountsHistogram {
     pub(crate) deleted_record_counts: Vec<i64>,
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) struct ProtocolMetadata {
+    pub(crate) protocol: Protocol,
+    pub(crate) metadata: Metadata,
+}
+
+impl ProtocolMetadata {
+    /// Attempt to read Protocol and Metadata from a CRC file. This will return an error if
+    /// protocol/metadata cannot be found in the CRC file (specification requires both to be
+    /// present).
+    pub(crate) fn try_from_crc(path: FileMeta, engine: &dyn Engine) -> DeltaResult<Self> {
+        let json = engine.json_handler();
+        let crc_schema = Crc::to_schema().into();
+
+        let mut crc_data = json.read_json_files(&[path], crc_schema, None)?;
+        let crc_batch = crc_data.next().ok_or(Error::generic(
+            "CRC file should contain exactly one JSON object (and therefore one batch)",
+        ))?;
+
+        let mut visitor = CrcProtocolMetadataVisitor::default();
+        visitor.visit_rows_of(crc_batch?.as_ref())?;
+        // note that if protocol/metadata are missing the visitor we return an error here
+        Ok(Self {
+            protocol: visitor
+                .protocol
+                .ok_or(Error::generic("Protocol not found in CRC file"))?,
+            metadata: visitor
+                .metadata
+                .ok_or(Error::generic("Metadata not found in CRC file"))?,
+        })
+    }
+}
+
 /// For now we just define a visitor for Protocol and Metadata in CRC files since (for now) that's
 /// the only optimization we implement. Since CRC files can contain lots of other data, we have a
 /// specific visitor for only Protocol/Metadata here.
-#[allow(unused)] // TODO: remove after we read CRCs
 #[derive(Debug, Default)]
 pub(crate) struct CrcProtocolMetadataVisitor {
-    pub(crate) protocol: Protocol,
-    pub(crate) metadata: Metadata,
+    pub(crate) protocol: Option<Protocol>,
+    pub(crate) metadata: Option<Metadata>,
```

Comment:
this is fixing an incorrect usage of `Default` --> we want Default to just set up 'empty' P/M for us to fill in after visiting - not to generate some default P/M itself.